connectors.go

```diff
@@ -97,74 +86,66 @@ var (
 	ErrUnknownConnector = errors.New("unknown connector")
 )
 
-// New returns a new Connector. The signature is generic to facilitate more flexible caller setup
-// (e.g. constructing a new connector based on parsing a config file, whose exact params
-// aren't known until runtime). However, if you can use the API.New form, it's preferred,
-// since you get type safety and more readable code.
-func New(provider providers.Provider, opts map[string]any) (Connector, error) { //nolint:ireturn
-	switch provider {
-	case providers.Salesforce:
-		return newSalesforce(opts)
-	case providers.Hubspot:
-		return newHubspot(opts)
-	default:
-		return nil, fmt.Errorf("%w: %s", ErrUnknownConnector, provider)
-	}
-}
-
-// newSalesforce returns a new Salesforce Connector, by unwrapping the options and passing them to the Salesforce API.
-func newSalesforce(opts map[string]any) (Connector, error) { //nolint:ireturn
-	var options []salesforce.Option
-
-	c, valid := getParam[common.AuthenticatedHTTPClient](opts, "client")
-	if valid {
-		options = append(options, salesforce.WithAuthenticatedClient(c))
-	}
-
-	w, valid := getParam[string](opts, "workspace")
-	if valid {
-		options = append(options, salesforce.WithWorkspace(w))
-	}
-
-	return Salesforce.New(options...)
+// NewProxyConnector returns a new proxy connector for the given provider with the given build options.
+func NewProxyConnector(
+	provider providers.Provider,
+	opts ...proxy.Option,
+) (*proxy.Connector, error) {
+	return proxy.NewConnector(provider, opts...)
 }
 
-// newHubspot returns a new Hubspot Connector, by unwrapping the options and passing them to the Hubspot API.
-func newHubspot(opts map[string]any) (Connector, error) { //nolint:ireturn
-	var options []hubspot.Option
-
-	c, valid := getParam[common.AuthenticatedHTTPClient](opts, "client")
-	if valid {
-		options = append(options, hubspot.WithAuthenticatedClient(c))
+// NewReadConnector returns a new read connector for the given provider with the given build options.
+func NewReadConnector(
```

Comment:
I like the interfaces approach. But I don't like New<interface_name>.

Generally speaking (not always) interfaces are provided for those who *use* the interface, not for those who *construct* the object. For example:

Bad:

```go
type MyInterface interface {
    Foo()
}

type MyImpl struct {}

func (m *MyImpl) Foo() { }

func (m *MyImpl) Bar() { } // CAN'T use this

func NewMyInterface() MyInterface {
    return &MyImpl{}
}

func UseMyInterface(i MyInterface) {
    i.Foo()
}

func main() {
    inst := NewMyInterface()
    UseMyInterface(inst)

    inst.Bar() // won't work
}
```

Good:

```go
type MyInterface interface {
    Foo()
}

type MyImpl struct {}

func (m *MyImpl) Foo() { }

func (m *MyImpl) Bar() { } // CAN use this

func UseMyInterface(i MyInterface) {
    i.Foo()
}

func main() {
    inst := &MyImpl{}
    UseMyInterface(inst)

    inst.Bar() // works
}
```

The distinction may seem academic but it's actually pretty important. A single object (er, struct) can implement multiple interfaces. And the interfaces may or may not have overlap with one another. So if you use the concrete type (for construction), the object is allowed to bring the full set of capabilities with it, rather than being whittled down to a singular purpose.

If you're still confused I'd recommend this article which explains it better than I did: https://duncanleung.com/go-idiom-accept-interfaces-return-types/

Note that I only loosely follow this idiom when I code. However in this case, because this is so central to the connectors repo, I think it's important.

Reply:
To put it in more ampersand-specific terms, many of our deep connectors will implement at least ProxyConnector, ReadConnector, and WriteConnector (probably others too). I don't think it makes sense that I have to construct 3 separate objects (using the 3 New functions) when a single object (using its original concrete type) can still do all these jobs.